### PR TITLE
EVG-7613 handle backend-not-reachable errors

### PR DESCRIPTION
**Pull request by ybrill**
If an error occurs reaching the backend (e.g.code-server isn't running) the default behavior is to return a 502 status code and an empty response. This will catch that error and pass a message and 500 status code.

Goes together with evergreen-ci/gimlet#77. Once that PR is approved it will be revendored and included here.

### CODE REVIEW

File: service/host.go [hunk 1]
```diff
@@ -323,3 +323,9 @@ func (uis *UIServer) getHostFromCache(hostID string) (*hostCacheItem, error) {
 
 	return &h, nil
 }
+
+func (uis *UIServer) handleBackendError(message string, statusCode int) func(http.ResponseWriter, *http.Request, error) {
+	return func(w http.ResponseWriter, r *http.Request, err error) {
+		gimlet.WriteTextResponse(w, statusCode, message)
```
Comment (tychoish) [inline service/host.go]: do we have the infrastructure to do the popflash thing? I suppose not. it'd be cool if there was a way to freeze the IDE until it came back.

File: service/ui.go [hunk 2]
```diff
@@ -358,6 +358,7 @@ func (uis *UIServer) GetServiceApp() *gimlet.APIApp {
 		FindTarget:        uis.getHostDNS,
 		StripSourcePrefix: true,
 		RemoteScheme:      "http",
+		ErrorHandler:      uis.handleBackendError("IDE service is not available", http.StatusInternalServerError),
```
Comment (tychoish) [inline service/ui.go]: is there some call to action that we can give users to take to remedy this situation? 

can we put retry logic in the proxy for blips/restarts of up-to 10s?

File: service/host.go (hunk 1 above)
Comment (ybrill) [reply service/host.go]: what do you mean by popflash?

File: service/ui.go (hunk 2 above)
Comment (ybrill) [reply service/ui.go]: I added a line to the response to check the code-server service is running.

I added a custom Transport to the ReverseProxy (in evergreen-ci/gimlet#78) and passed in a rehttp RoundTripper.

File: service/host.go (hunk 1 above)
Comment (tychoish) [reply service/host.go]: the thing that the legacy uses to communicate errors on pages, I think the function is called PopFlash or Flash

**Pull request merged**